llvm/lib/Target/RISCV/RISCVExpandAtomicPseudoInsts.cpp

```diff
@@ -1085,13 +1084,34 @@ bool RISCVExpandAtomicPseudo::expandAtomicCmpXchgCap(
   // .looptail:
   //   sc.c scratch, newval, (addr)
   //   bnez scratch, loophead
-  BuildMI(LoopTailMBB, DL,
-          TII->get(getSCForRMWCap(PtrIsCap, ExplicitAddrMode, Ordering, CLen)),
-          ScratchReg)
-      .addReg(AddrReg)
-      .addReg(NewValReg);
+  const MCInstrDesc &SCInst =
+      TII->get(getSCForRMWCap(PtrIsCap, ExplicitAddrMode, Ordering, CLen));
+  Register SCResultReg = ScratchReg;
+  if (ExplicitAddrMode) {
+    assert(RISCV::GPCRRegClass.contains(ScratchReg));
+    // With an explicit cap/ddc-relative SC, the register stored is also the
+    // success/failure writeback register.
+    // .looptail:
+    //   mv scratch, newval
+    //   sc.cap.c scratch, (addr)  (implicit rd=scratch)
+    //   bnez scratch, loophead
+    BuildMI(LoopTailMBB, DL, TII->get(RISCV::CMove), ScratchReg)
+        .addReg(NewValReg);
+    // Note: SC_C_CAP has the address register as the second argument not the
+    // first even though it is called rs1 in tablegen.
```

Comment:
That should just be fixed tbh; I hadn't realised that the normal SC/AMOs had operands in the normal order unlike plain stores which flip them like this.

Reply:
Should be fixed now.